File: src/p297/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import collections


class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if root is None:
            return '[]'
        l = []
        q = collections.deque()
        q.append(root)
        while len(q) != 0:
            node = q.popleft()
            l.append(None if node is None else node.val)
            if node is None:
                continue
            q.append(node.left)
            q.append(node.right)
        return str(l)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        l = [x.strip() for x in data[1:-1].split(',') if x.strip()]
        if not l:
            return None
        root = TreeNode(int(l[0]))
        q = collections.deque()
        q.append(root)
        pos = 1
        while len(q) != 0:
            node = q.popleft()
            if pos >= len(l) or l[pos] == 'None':
                node.left = None
            else:
                node.left = TreeNode(int(l[pos]))
                q.append(node.left)
            pos += 1
            if pos >= len(l) or l[pos] == 'None':
                node.right = None
            else:
                node.right = TreeNode(int(l[pos]))
                q.append(node.right)
            pos += 1
        return root


#Your Codec object will be instantiated and called as such:
# root = TreeNode(1)
# root.right = TreeNode(2)
# codec = Codec()
# data = codec.serialize(root)
# print(data)
# codec.deserialize(data)

File: src/p297/solution.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    string to(int v) {
        stringstream ss;
        ss << v;
        return ss.str();
    }
    int to(string s) {
        stringstream ss(s);
        int v;
        ss >> v;
        return v;
    }
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "";
        string result = to(root->val);
        if (root->left == NULL && root->right == NULL) return result;
        string left = serialize(root->left), right = serialize(root->right);
        result = result + "(" + left + ")" + "(" + right + ")";
        return result;    
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data == "") return NULL;
        int left_start=-1, left_end=-1, right_start=-1, right_end=-1, level = 0;
        for (int i = 0 ; i < data.length(); i++) {
            if (data[i] == '(') {
                if (level == 0) {
                    if (left_start == -1) left_start = i;
                    else right_start = i;
                }
                level++;
            }
            if (data[i] == ')') {
                level--;
                if (level == 0) {
                    if (left_end == -1) left_end = i;
                    else right_end = i;
                }
            }
        }
        int val = (left_start == -1 ? to(data) : to(data.substr(0, left_start)));
        TreeNode* node = new TreeNode(val);
        if (left_start != -1) {
            node->left = deserialize(data.substr(left_start + 1, left_end - left_start - 1));
            node->right = deserialize(data.substr(right_start + 1, right_end - right_start - 1));
        }
        return node;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));
